include/boost/geometry/algorithms/for_each.hpp

```diff
@@ -94,31 +96,76 @@ struct fe_range_per_point
 };
 
 
+template
+<
+    typename Range,
+    typename Point = typename add_const_if_c
+        <
+            boost::is_const<Range>::value,
+            typename point_type<Range>::type
+        >::type,
+    bool UseRefernces =
+        ( boost::is_const<Range>::value
+       || boost::is_same
+            <
+                typename boost::range_reference<Range>::type,
+                Point&
+            >::value )
+>
+struct fe_range_per_segment_call_f
+{
+    template <typename Iterator, typename Functor>
+    static inline void apply(Iterator it0, Iterator it1, Functor& f)
+    {
+        // Implementation for true references (both const and mutable)
+        // and const non-true references.
+        // If non-true const references are returned by iterators
+        // then const true-references here prevents temporary
+        // objects from being destroyed.
+        Point& p0 = *it0;
+        Point& p1 = *it1;
+        model::referring_segment<Point> s(p0, p1);
+        f(s);
+    }
+};
+
+template <typename Range, typename Point>
+struct fe_range_per_segment_call_f<Range, Point, false>
+{
+    template <typename Iterator, typename Functor>
+    static inline void apply(Iterator it0, Iterator it1, Functor& f)
+    {
+        // Non-true mutable references returned by iterators
+        // temporary points has to be created and assigned afterwards.
+        Point p0 = *it0;
+        Point p1 = *it1;
+        model::referring_segment<Point> s(p0, p1);
```

Comment:
Clear, it's fine to me. I kind of see the problem. Is it ever used like this? But anyway, it supports what it did before. Thank you.

(You can change `has` to `have`)

Reply:
I think using `for_each_segment` for mutable ranges doesn't make sense but since they were supported I decided to keep that behavior even for proxy references. We could talk about dropping it, in fact I have a next PR nearly ready related to for_each so we could talk about it there because this PR is fixing a bug.